libraries/AP_OSD/AP_OSD_ParamScreen.cpp

```diff
@@ -0,0 +1,663 @@
+void AP_OSD_ParamScreen::save_parameters()
+{
+    if (!_requires_save) {
+        return;
+    }
+
+    for (uint8_t i = 0; i < NUM_PARAMS; i++) {
+        if (params[i].enabled) {
+            AP_Param* p = params[i]._param;
+            if (p != nullptr) {
+                p->save();
+            }
+            params[i].save_as_new();
+        }
+    }
+    _requires_save = false;
+}
+
+// return radio values as LOW, MIDDLE, HIGH
+RC_Channel::AuxSwitchPos AP_OSD_ParamScreen::get_channel_pos(uint8_t rcmapchan) const
+{
+    const RC_Channel* chan = rc().channel(rcmapchan-1);
+    if (chan == nullptr) {
+        return RC_Channel::AuxSwitchPos::LOW;
+    }
+
+    const uint16_t in = chan->get_radio_in();
+    if (in <= 900 or in >= 2200) {
+        return RC_Channel::AuxSwitchPos::LOW;
+    }
+
+    // switch is reversed if 'reversed' option set on channel and switches reverse is allowed by RC_OPTIONS
+    bool switch_reversed = chan->get_reverse() && rc().switch_reverse_allowed();
+
+    if (in < 1300) {
+        return switch_reversed ? RC_Channel::AuxSwitchPos::HIGH : RC_Channel::AuxSwitchPos::LOW;
+    } else if (in > 1700) {
+        return switch_reversed ? RC_Channel::AuxSwitchPos::LOW : RC_Channel::AuxSwitchPos::HIGH;
+    } else {
+        return RC_Channel::AuxSwitchPos::MIDDLE;
+    }
+}
+
+// map rc input to an event
+AP_OSD_ParamScreen::Event AP_OSD_ParamScreen::map_rc_input_to_event() const
+{
+    const RC_Channel::AuxSwitchPos throttle = get_channel_pos(AP::rcmap()->throttle());
+    const RC_Channel::AuxSwitchPos yaw = get_channel_pos(AP::rcmap()->yaw());
+    const RC_Channel::AuxSwitchPos roll = get_channel_pos(AP::rcmap()->roll());
+    const RC_Channel::AuxSwitchPos pitch = get_channel_pos(AP::rcmap()->pitch());
+
+    Event result = Event::NONE;
+
+    if (yaw != RC_Channel::AuxSwitchPos::MIDDLE || throttle != RC_Channel::AuxSwitchPos::LOW) {
+        return result;
+    }
+
+    if (pitch == RC_Channel::AuxSwitchPos::MIDDLE && roll == RC_Channel::AuxSwitchPos::LOW) {
+        result = Event::MENU_EXIT;
+    } else if (pitch == RC_Channel::AuxSwitchPos::MIDDLE && roll == RC_Channel::AuxSwitchPos::HIGH) {
+        result = Event::MENU_ENTER;
+    } else if (pitch == RC_Channel::AuxSwitchPos::LOW && roll == RC_Channel::AuxSwitchPos::MIDDLE) {
+        result = Event::MENU_UP;
+    } else if (pitch == RC_Channel::AuxSwitchPos::HIGH && roll == RC_Channel::AuxSwitchPos::MIDDLE) {
+        result = Event::MENU_DOWN;
+    } else {
+        // OSD option has not changed so assume stick re-centering
+        result = Event::NONE;
+    }
+    return result;
+}
+
+// update the state machine when disarmed
+void AP_OSD_ParamScreen::update_state_machine()
+{
+    const uint32_t now = AP_HAL::millis();
+    if ((now - _transition_start_ms) < _transition_timeout_ms) {
+        return;
+    }
+
+    const Event ev = map_rc_input_to_event();
+    // only take action on transitions
+    if (ev == Event::NONE && ev == _last_rc_event) {
```

Comment:
This looks... odd....

Reply:
You only skip this step when you get None twice in a row.